NEWS

```diff
@@ -3,6 +3,21 @@
 History
 -------
 
+1.5.0 (2018-06-22)
+~~~~~~~~~~~~~~~~~~
+
+* New ``Selector.attrib`` and ``SelectorList.attrib`` properties which make
+  it easier to get attributes of HTML elements.
+* CSS selectors became faster: compilation results are cached
+  (LRU cache is used for ``css2xpath``), so there is
+  less overhead when the same CSS expression is used several times.
+* ``.get()`` and ``.getall()`` selector methods are documented and recommended
+  over ``.extract_first()`` and ``.extract()``.
+* ``.extract()`` and  ``.extract_first()`` are now implemented using
+  ``.get()`` and ``.getall()``, not the other way around. It can be
+  **backwards incompatible** in case of custom Selector subclasses.
```

Comment:
this is not a good idea then, we can do better by providing proper methods instead of aliases for `extract` and `extract_first`. I am sure there are projects extending Selector class out there.

Reply:
Hey @kmike, in which case exactly this would be backwards incompatible?

Reply:
hey! So, we used to have code like this (simplified):

```py
class Selector:
    def extract(self):
        # ...
    get = extract

class SelectorList:
    def extract(self):
        return [s.extract() for s in self]
    getall = extract
```

In https://github.com/scrapy/parsel/pull/113/commits/5739a929fe7d124f8faafea8a4ffd4ff4e95f20d it is changed to this:

```py
class Selector:
    def get(self):
        # ...
    extract = get

class SelectorList:
    def getall(self):
        return [s.get() for s in self]
    extract = getall
```

First, there is an issue with aliases: if one overrides e.g. Selector.extract in a subclass, then Selector.get is not changed automatically, alias needs to be re-added in a subclass. It is not necessarily a problem, and it saves us a function call; also, nothing changed regarding this in the "blessing" PRs.

If we make one of these methods call another, instead of having aliases, then overriding would work as expected - but only for one of the methods, and we're to choose for which method. I'm not sure it is a big win.

The backwards incompatible change here is a change in implementation of methods: getall used to call .extract, while now it calls .get; there are several other changes like this. It means that Selector.extract override in a subclass no longer has an effect on SelectorList, while .get override starts to have an effect.

I made this change while fixing a docstring for .getall. It said that getall calls .extract methods of Selector, not .get, which I was thinking is not good. 

So, an exact backwards incompatibility is that in custom Selector subclasses .get should be overridden instead of .extract to take an effect in `SelectorList.getall`, `SelectorList.extract`, `SelecrorList.getall`, `SelectorList.extract_first`, `Selector.__repr__`, `Selector.re`, `Selector.bool`. It is not a problem as soon as Selector subclass sets up aliases properly, i.e. it is only an issue when a subclass is overriding .extract without fixing .get alias.

Reply:
@dangra do you think it can be resolved by documenting the backwards incompatibility better, or if something should be changed code-wise?

Reply:
So if I understood correctly, of we delegate at method level, like:

```
def getall(self):
    return self.extract()
```

... we'd avoid the incompatibility with subclasses that override `extract` and use super, right?

Reply:
@eliasdorneles unfortunately nope, the problem is not with these aliases, the problem is that `SelectorList.extract` now calls `Selector.get` instead of `Selector.extract`.

Reply:
The issue I'm talking about is not SelectorList subclasses which override some methods (nothing changed here since parsel 1.2, though your and Daniel's suggestion may help), but about Selector subclasses.

```
class MySelector(Selector):
    def get(self):
        return super().get() + "hello!"

MySelector(response.text).css('a::text').getall()  # it should use .get of a Selector, right?
MySelector(response.text).css('a::text').extract()  # it should use .get of a Selector too, right?
```

One option is to have two "parallel" set of methods: extract / extract_first use extract, get/getall use get. It is still not a complete solution, as there is also re, repr and bool methods. It'd also be an accumulated cruft.

Reply:
Go for documentation then ! :)

Reply:
@dangra done!